components/base/sensorcontrolled/movestraight.go

```diff
@@ -0,0 +1,204 @@
+			return err
+		}
+	}
+
+	// initialize relevant parameters for moving straight
+	slowDownDist := calcSlowDownDist(distanceMm)
+
+	var initPos *geo.Point
+
+	if sb.position != nil {
+		initPos, _, err = sb.position.Position(ctx, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	ticker := time.NewTicker(time.Duration(1000./sb.controlLoopConfig.Frequency) * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			// do not return context canceled errors, just log them
+			if errors.Is(ctx.Err(), context.Canceled) {
+				sb.logger.Error(ctx.Err())
+				return nil
+			}
+			return ctx.Err()
+		case <-ticker.C:
+			var errDist float64
+
+			angVelDes, err := sb.calcHeadingControl(ctx, initialHeading)
+			if err != nil {
+				return err
+			}
+
+			if sb.position != nil {
+				errDist, err = sb.calcPositionError(ctx, distanceMm, initPos)
+				if err != nil {
+					return err
+				}
+			}
+
+			if errDist < moveStraightErrTarget {
+				return sb.Stop(ctx, nil)
+			}
+
+			linVelDes := calcLinVel(errDist, mmPerSec, slowDownDist)
+			if err != nil {
+				return err
+			}
+
+			// update velocity controller
+			if err := sb.updateControlConfig(ctx, linVelDes/1000.0, angVelDes); err != nil {
+				return err
+			}
+
+			// exit if the straight takes too long
+			if time.Since(startTime) > timeOut {
+				sb.logger.CWarn(ctx, "exceeded time for MoveStraightCall, stopping base")
+
+				return sb.Stop(ctx, nil)
+			}
+		}
+	}
+}
+
+// calculate the desired angular velocity to correct the heading of the base.
+func (sb *sensorBase) calcHeadingControl(ctx context.Context, initHeading float64) (float64, error) {
+	currHeading, err := sb.headingFunc(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	headingErr := initHeading - currHeading
+	headingErrWrapped := headingErr - (math.Floor((headingErr+180.)/(2*180.)))*2*180. // [-180;180)
+
+	return headingErrWrapped * headingGain, nil
+}
+
+// calcPositionError calculates the current error in position.
+// This results in the distance the base needs to travel to reach the goal.
+func (sb *sensorBase) calcPositionError(ctx context.Context, distanceMm int, initPos *geo.Point) (float64, error) {
+	pos, _, err := sb.position.Position(ctx, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	// the currDist will always return as positive, so we need the goal distanceMm to be positive
+	currDist := initPos.GreatCircleDistance(pos) * 1000000.
+	return math.Abs(float64(distanceMm)) - currDist, nil
+}
+
+// calcLinVel computes the desired linear velocity based on how far the base is from reaching the goal.
+func calcLinVel(errDist, mmPerSec, slowDownDist float64) float64 {
+	// have the velocity slow down when appoaching the goal. Otherwise use the desired velocity
+	linVel := errDist * mmPerSec / slowDownDist
+	absMmPerSec := math.Abs(mmPerSec)
+	if linVel > 0 && linVel > absMmPerSec {
+		return absMmPerSec
+	}
+	if linVel < 0 && linVel < -absMmPerSec {
+		return -absMmPerSec
+	}
```

Comment:
this logic ended up being kinda gross. based on the direction of the velocity we need to set the upper bound of the desired velocity